### ISSUE TITLE
Launch target is not selectable when executable is installed in custom folder

### ISSUE THREAD

**Issue by prajapak** [microsoft/vscode-makefile-tools#380]
Can't select current launch target if it is installed in custom folder. It points to default target on workspace directory. For example,

1. Makefile installs executable in ${workspace}/build/executable/debug/ 
2. Now when click edit in "Launch target: [unset], it pops up ${workspace}>a.out() which doesn't exists in fact.
3. Current launch target is defined in settings.json which is ${workspace}/build/executable/debug/HelloWorld

Here is my settings.json,

```
{
      "makefile.launchConfigurations": [
       {
            "cwd": "${workspaceFolder}",
            "binaryPath": "${workspaceFolder}/build/executable/debug/HelloWorld",
            "binaryArgs": [],
            "miDebuggerPath": "/usr/bin/gdb",
            "MIMode": "gdb"
        }
    ],
    "makefile.buildBeforeLaunch": true
}
```

Prompt help will be much appreciated.

**Comment by bobbrow** [comment]
And if you are able to select HelloWorld as the active target in the sidebar, you could substitute the hard-coded path with `${command:makefile.launchTargetPath}`

**Comment by prajapak** [comment]
Bob,Using launch.json works fine. Question is to do from make file menu.Regards,AshokOn Nov 7, 2022, at 12:39 PM, Bob Brown ***@***.***> wrote:﻿
Sorry for the delay in responding.  My engineer who knows this best has been on leave.  It sounds like there may be a bug here.  The launchConfigurations are intended to be a simplification of launch.json.  If it's not working as expected, you can try using VS Code's launch.json instead and press F5 to start debugging.
An example to get you started would be something like this:
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Debug the program",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/build/executable/debug/HelloWorld",
            "cwd": "${workspaceFolder}",
            "miDebuggerPath": "/usr/bin/gdb",
            "MIMode": "gdb"
        }
    ]
}

**Comment by bobbrow** [comment]
I understand. I was providing a workaround until @andreeis comes back and can help investigate.

**Comment by prajapak** [comment]
Thanks Bob. I understand. As you said, it seems a bug.Regards,AshokOn Nov 7, 2022, at 2:11 PM, Bob Brown ***@***.***> wrote:﻿
I understand. I was providing a workaround until @andreeis comes back and can help investigate.

**Comment by prajapak** [comment]
Hi Bob,Is there any update on this?Regards,AshokOn Nov 7, 2022, at 2:11 PM, Bob Brown ***@***.***> wrote:﻿
I understand. I was providing a workaround until @andreeis comes back and can help investigate.

**Comment by bobbrow** [comment]
Sorry, no update yet.  @andreeis will take a look when she returns.

**Comment by prajapak** [comment]
Any idea when she returns?Thanks Bob.Regards,AshokOn Nov 15, 2022, at 4:11 PM, Bob Brown ***@***.***> wrote:﻿
Sorry, no update yet.  @andreeis will take a look when she returns.

**Comment by andreeis** [comment]
@prajapak, I am back, at least to address all the recent feedback on GitHub. Very likely there will still be some OOF but less and less. I'll be back with more details after I digest all info on this thread. Thank you for your patience.

**Comment by prajapak** [comment]
Thanks Andrea. Let me know if you need more info.Regards,AshokOn Nov 16, 2022, at 12:46 PM, Andreea Isac ***@***.***> wrote:﻿
@prajapak, I am back, at least to address all the recent feedback on GitHub. Very likely there will still be some OOF but less and less. I'll be back with more details after I digest all info on this thread. Thank you for your patience.

**Comment by andreeis** [comment]
@prajapak, besides making sure that the binary exists on disk (we don't show it in the quickPick if it's missing, which can also be debatable if we want or not to display it) I was able to reproduce this problem when I use the ${workspaceFolder} variable. We don't expand it when reading the launch configurations. We'll fix it for 0.7. In the meantime, use full paths and you should see it in the quickPick. Let me know if not.

Also, maybe less important for you, "a.out": probably that is included because we see a linker command that does not list an explicit binary and we assume the "a.out" on a linux system. If that looks wrong to you you can share us some logs and we can inspect the command line that makes us think there is an "a.out" as a launch target. Although indeed for that (the launch targets we deduce/assume we don't check if they exist on disk like we do for other launch targets that are given explicitly).

**Comment by prajapak** [comment]
Andrea,I am talking about make file configuration not launch target setting. Path expansion in makefile configuration did help. It works now.Thank you for help. Looking forward for updated version.Regards,AshokOn Nov 17, 2022, at 3:48 PM, Andreea Isac ***@***.***> wrote:﻿
@prajapak, besides making sure that the binary exists on disk (we don't show it in the quickPick if it's missing, which can also be debatable if we want or not to display it) I was able to reproduce this problem when I use the ${workspaceFolder} variable. We don't expand it when reading the launch configurations. We'll fix it for 0.7. In the meantime, use full paths and you should see it in the quickPick. Let me know if not.
Also, maybe less important for you, "a.out": probably that is included because we see a linker command that does not list an explicit binary and we assume the "a.out" on a linux system. If that looks wrong to you you can share us some logs and we can inspect the command line that makes us think there is an "a.out" as a launch target. Although indeed for that (the launch targets we deduce/assume we don't check if they exist on disk like we do for other launch targets that are given explicitly).

**Comment by prajapak** [comment]
Hi Andrea,Is there any tentative date for 0.7 release?Regards,AshokOn Nov 18, 2022, at 9:37 AM, Ashok Prajapati ***@***.***> wrote:﻿Andrea,I am talking about make file configuration not launch target setting. Path expansion in makefile configuration did help. It works now.Thank you for help. Looking forward for updated version.Regards,AshokOn Nov 17, 2022, at 3:48 PM, Andreea Isac ***@***.***> wrote:﻿
@prajapak, besides making sure that the binary exists on disk (we don't show it in the quickPick if it's missing, which can also be debatable if we want or not to display it) I was able to reproduce this problem when I use the ${workspaceFolder} variable. We don't expand it when reading the launch configurations. We'll fix it for 0.7. In the meantime, use full paths and you should see it in the quickPick. Let me know if not.
Also, maybe less important for you, "a.out": probably that is included because we see a linker command that does not list an explicit binary and we assume the "a.out" on a linux system. If that looks wrong to you you can share us some logs and we can inspect the command line that makes us think there is an "a.out" as a launch target. Although indeed for that (the launch targets we deduce/assume we don't check if they exist on disk like we do for other launch targets that are given explicitly).

**Comment by andreeis** [comment]
@prajapak 
While we wrap up the final testing for the 0.7.0 release, if you would like you can preview the variable expansion feature if you download [this vsix](https://github.com/microsoft/vscode-makefile-tools/actions/runs/4703905680) and install it via the "Install from VSIX" command in the palette. 

Besides the highly requested ${workspaceFolder} and ${workspaceRoot} macros that work now from any setting entry-point, we support also:
- ${userHome}
- ${env:ENVIRONMENT_VARIABLE}
- ${config:any_extension_scope.any_setting_id}
- ${command:any_extension_scope.any_command_name} (this one worked already in launch and tasks json but now they work in settings.json as well). Makefile Tools has the same launch commands we implemented for a while already and you can use any other command that is available. If it returns something useful, we'll expand with the return value.
- ${configuration} (variable syntax) and ${command:makefile.getConfiguration} (command syntax, accessible in launch tasks json as well) both return what is the current active configuration (what you selected in the left UI, either one of the entries in makefile.configurations from settings.json or "Default")
- ${buildTarget} and ${command:makefile.getBuildTarget} map to what build target you selected in the UI (or via the command set build target), so it can be "all" or any other individual target from your makefile or expands to empty if you didn't select anything in the quickPick.

If by any chance you need to pass strings with the varexp syntax and you don't want us to expand it use \\ to escape it (maybe you pass ${something} via makeArgs or binaryArgs). If that happens in a path, use forward slash for path separator so that we don't confuse it with the escape character.

If we don't recognize the pattern within ${} or something happens while running the referenced command or... we expand to "unknown".
If the first expansion resolves to another ${} varexp syntax we complain in the log that we don't support yet multiple expansion passes and the result is still "unknown". We will support more complex scenarios in future.

Any feedback?

**Comment by prajapak** [comment]
Hi Andrea,Thanks for update. I will check it out and let you know.Regards,AshokOn Apr 14, 2023, at 5:45 PM, Andreea Isac ***@***.***> wrote:﻿
@prajapak
While we wrap up the final testing for the 0.7.0 release, if you would like you can preview the variable expansion feature if you download this vsix and install it via the "Install from VSIX" command in the palette.
Besides the highly requested ${workspaceFolder} and ${workspaceRoot} macros that work now from any setting entry-point, we support also:

${userHome}
${env:ENVIRONMENT_VARIABLE}
${config:any_extension_scope.any_setting_id}
${command:any_extension_scope.any_command_name} (this one worked already in launch and tasks json but now they work in settings.json as well). Makefile Tools has the same launch commands we implemented for a while already and you can use any other command that is available. If it returns something useful, we'll expand with the return value.
${configuration} (variable syntax) and ${command:makefile.getConfiguration} (command syntax, accessible in launch tasks json as well) both return what is the current active configuration (what you selected in the left UI, either one of the entries in makefile.configurations from settings.json or "Default")
${buildTarget} and ${command:makefile.getBuildTarget} map to what build target you selected in the UI (or via the command set build target), so it can be "all" or any other individual target from your makefile or expands to empty if you didn't select anything in the quickPick.

If by any chance you need to pass strings with the varexp syntax and you don't want us to expand it use \ to escape it (maybe you pass ${something} via makeArgs or binaryArgs). If that happens in a path, use forward slash for path separator so that we don't confuse it with the escape character.
If we don't recognize the pattern within ${} or something happens while running the referenced command or... we expand to "unknown".
If the first expansion resolves to another ${} varexp syntax we complain in the log that we don't support yet multiple expansion passes and the result is still "unknown". We will support more complex scenarios in future.
Any feedback?

**Comment by andreeis** [comment]
Closing this since it seems that all the issues that needed to be addressed were related to variable expansion. @prajapak , if there is any other scenario that I missed while reviewing this thread and which is not related to variable expansion (released in 0.7) reactivate or create a new issue for us to look at.

**Comment by prajapak** [comment]
Sorry for trying it late. I am using latest version of MakeFike ext and it seems to have same issue. It didn’t work until I provide absolute path aka it did not expand $WorkSpaceFolder.
